File: LeetCode/2301-2400/Problem2326_SpiralMatrixIV/SpiralMatrixIV.py
class Solution:
    def spiralMatrix(self, m: int, n: int, head: Optional[ListNode]) -> List[List[int]]:
        matrix = [[-1 for _ in range(n)] for _ in range(m)]
        
        top, bottom = 0, m - 1
        left, right = 0, n - 1

        while head and top <= bottom and left <= right:
            for j in range(left, right + 1):
                if not head:
                    return matrix
                matrix[top][j] = head.val
                head = head.next
            top += 1

            for i in range(top, bottom + 1):
                if not head:
                    return matrix
                matrix[i][right] = head.val
                head = head.next
            right -= 1

            for j in range(right, left - 1, -1):
                if not head:
                    return matrix
                matrix[bottom][j] = head.val
                head = head.next
            bottom -= 1

            for i in range(bottom, top - 1, -1):
                if not head:
                    return matrix
                matrix[i][left] = head.val
                head = head.next
            left += 1

        return matrix

File: LeetCode/2301-2400/Problem2326_SpiralMatrixIV/SpiralMatrixIV.cpp
class Solution {
public:
    vector<vector<int>> spiralMatrix(int m, int n, ListNode* head) {
        vector<vector<int>> matrix(m, vector<int>(n, -1));
        
        int top = 0, bottom = m - 1;
        int left = 0, right = n - 1;
        
        while (head != nullptr && top <= bottom && left <= right) {
            for (int j = left; j <= right && head != nullptr; ++j) {
                matrix[top][j] = head->val;
                head = head->next;
            }
            top++;
            
            for (int i = top; i <= bottom && head != nullptr; ++i) {
                matrix[i][right] = head->val;
                head = head->next;
            }
            right--;
            
            for (int j = right; j >= left && head != nullptr; --j) {
                matrix[bottom][j] = head->val;
                head = head->next;
            }
            bottom--;
            
            for (int i = bottom; i >= top && head != nullptr; --i) {
                matrix[i][left] = head->val;
                head = head->next;
            }
            left++;
        }
        
        return matrix;
    }
};